### PR TITLE
Limit subscribers

### PR DESCRIPTION
**Pull request by edsko**
This PR looks bigger than it is because it depends on the not-yet-merged https://github.com/input-output-hk/cardano-sl/pull/1286 .

### CODE REVIEW

**Comment by edsko** [comment]
@dcoutts @avieth this is work in progress, but one thing I noticed while I was working on this: for P2P node we had

``` haskell
topologySubscriberNodeType :: Topology kademlia -> Maybe NodeType
topologySubscriberNodeType TopologyP2P{}         = Just NodeRelay
...
```

That's wrong, isn't it? Relays don't subscribe. Other P2P nodes may subscribe to P2P nodes, but we assume that they must be edge nodes, no? I've changed it to

``` haskell
topologySubscribers :: Topology kademlia -> Maybe (NodeType, OQ.MaxBucketSize)
topologySubscribers TopologyP2P{..}         = Just (NodeEdge, topologyMaxSubscrs)
...
```

**Comment by edsko** [comment]
I will continue with this probably tomorrow or later today, but will rebase it once https://github.com/input-output-hk/cardano-sl/pull/1347 has been merged.

**Comment by avieth** [comment]
RE p2p subscriber type: I'm not sure. I can never remember, because both of them make sense. But I'd imagine `NodeRelay` is the right choice, since every p2p member is responsible for relaying throughout the kademlia network.

**Comment by edsko** [comment]
> RE p2p subscriber type: I'm not sure. I can never remember, because both of them make sense. But I'd imagine NodeRelay is the right choice, since every p2p member is responsible for relaying throughout the kademlia network.

P2P nodes are assumed to be edge nodes, so I suspect `NodeEdge` is the correct one :)

**Comment by edsko** [comment]
@dcoutts ready for review. Most of this is just a bit of code cleanup; important changes:

* Added `maxSubscrs` fields to the various relevant topology `.yaml` formats
* In the subscription listener, check if the bucket size got exceeded and if so, close the conversation immediately, effectively indicating to the subscriber that subscription failed
* In the worker we already assumed that if the conversation got closed the subscription failed and we already would attempt a different one, so that didn't need to change.

We do still need to deal with valency and fallbacks; I'll do that in a separate PR.

**Comment by edsko** [comment]
Rebased on master.

**Comment by edsko** [comment]
@dcoutts did you consider my earlier question (https://github.com/input-output-hk/cardano-sl/pull/1354#issuecomment-322772039)? If you agree with me, there's nothing to do, but if not, we should fix it :)

File: infra/Pos/Subscription/Common.hs [hunk 1]
```diff
@@ -86,9 +86,12 @@ subscriptionListener nodeType = listenerConv @Void $ \__ourVerInfo nodeId conv -
...
       let peers = simplePeers [(nodeType, nodeId)]
-      bracket_ (updatePeersBucket BucketSubscriptionListener (<> peers))
-               (updatePeersBucket BucketSubscriptionListener (removePeer nodeId))
-               (void $ recvLimited conv)
+      bracket
+        (updatePeersBucket BucketSubscriptionListener (<> peers))
+        (\added -> when added $
+          void $ updatePeersBucket BucketSubscriptionListener (removePeer nodeId))
+        (\added -> when added $
+          void $ recvLimited conv) -- if not added, close the conversation
```
Comment (avieth) [inline infra/Pos/Subscription/Common.hs]: Perhaps we should put some logging in here, particularly for the case where a subscriber can't be added.

Comment (edsko) [inline infra/Pos/Subscription/Common.hs]: Note that the queue will already log this, and exposes this as an EKG counter as well.

Comment (avieth) [inline infra/Pos/Subscription/Common.hs]: Oh alright, don't bother logging here then.

**Comment by edsko** [comment]
> AFAICT max bucket size isn't mutable and SIGHUP doesn't actually change it. Planning to implement that in this PR, or later?

Correct. Discussed that with @dcoutts and decided to leave it for now. Later PR.

**Comment by dcoutts** [comment]
> @dcoutts did you consider my earlier question (#1354 (comment))? If you agree with me, there's nothing to do, but if not, we should fix it :)

Thanks for the reminder on this detail.

> Other P2P nodes may subscribe to P2P nodes, but we assume that they must be edge nodes, no?

So of course we've not completely finished the p2p case, but we're assuming that when we do that p2p nodes will subscribe to each other. So what should p2p nodes believe the the node type of their peers to be? Recall that in our current policy edge nodes only talk to relay nodes and not to other edge nodes! So currently the answer must in fact be relay, and if we wanted to change that then we'd have to extend the policy to allow edge nodes to talk to edge nodes.

But if we look at the edge -> relay policy, it should actually fit the p2p case just fine:

```
Enqueing:
 * header announce : to Relay, P1, 3 route list, max ahead 0, fallback drop
 * header request  : to Relay, P2, 3 route list, max ahead 1, fallback drop
 * block  request  : to Relay, P2, one of,       max ahead 1, fallback TODO
 * transaction     : to Relay, P4, 3 route list, max ahead 3, fallback drop
 * mpc             : to Relay not sent!
```
So we'll send everything including block headers and transactions, but not MPC.

As an alternative we could say that p2p edge nodes believe their peers to be edge nodes but then they'd still need to have the above as their policy for edge -> edge nodes.

I don't mind either way, perhaps it's less confusing to say they're all edge and adjust the policy for p2p nodes so that they never talk to relays and only to what they believe to be edge nodes. If they discover an actual relay node via kademlia, they'd consider it the same as any other "real" p2p edge node.

```
topologySubscribers :: Topology kademlia -> Maybe (NodeType, OQ.MaxBucketSize)
topologySubscribers TopologyP2P{..}         = Just (NodeEdge, topologyMaxSubscrs)
```

So for the moment we should switch it back to `NodeRelay`, but we can do the conversion above to switch it consistently over to edge as a later PR.

**Comment by avieth** [comment]
I took the liberty of rebasing to fix those conflicts and I'm going to force push @edsko beware!

**Comment by edsko** [comment]
>  Recall that in our current policy edge nodes only talk to relay nodes and not to other edge nodes! So currently the answer must in fact be relay,

Ah, yes, quite right. This is confusing :) But ok, makes sense. Apologies for doubting you @avieth :)

**Pull request merged**